### PR TITLE
Notification prototype

### PR DESCRIPTION
**Pull request by EffortlessFury**
This is a prototype of a notification system. The goal of this PR is two-fold:

A. Discuss and validate the mechanism by which notifications are generated and fanned-out to notification displays.

B. Design the initial feature that will use the system.

For the demonstration of the mechanism, this is currently setup to send a notification on each refresh if there is new gear added to your inventory (first load will send a notification with all gear). I'm randomizing the order so that the main gear image and text is different each refresh.

The original impetus for this feature was to get notification when I received "good gear." The design allows anyone to design a service that:

- Generates notifications (The receipt of a good piece of gear, in this example, the receipt of any new gear, in this prototype, and in the end, I want to design a service that allows a user to define a filter expression (the ones we use for search) to dictate what to be notified about) 

- Handles displaying the notifications to a user (in this prototype, the browser displays the notification, but one service I plan to write is to shuttle the notifications to PushBullet). Another receiver of the notifications could be a feed in the UI to see a history of notifications that could also possibly support a more detailed view for each notification.

Other scenarios for when you want a notification might be when your inventory space is filling or you get an item with a specific perk, roll, etc.

_______

For the first feature, I'd like to design a notification generator that can notify a user when gear is received that matches a saved search. I'd also like to implement two notification displays: Browser window and PushBullet.

### CODE REVIEW

File: src/scripts/services/notifications/dimItemNotificationGenerator.factory.js [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    // Ideally we'd start/stop this as the user turns it on/off to avoid unneeded processing.
+    return {
+      active: false,
+      start: () => {
+        var self = this;
+        function checkForNewStuff(store) {
+          if (!store) {
+            return;
+          }
+          const newItems = shuffle(_.filter(store.items, function(item) { return !item.notransfer && item.equipment && !item.classified && _.indexOf(alreadyNotified, item.id) < 0; }));
```
Comment (kyleshay) [inline src/scripts/services/notifications/dimItemNotificationGenerator.factory.js]: really cool idea!

curious here, why shuffle? 

btw - we also already have a way to detect if an item is new, could we link into that?
https://github.com/DestinyItemManager/DIM/blob/dev/src/scripts/services/dimStoreService.factory.js#L872

**Comment by kyleshay** [comment]
Thanks @EffortlessFury ! whats the main goal of this system? Could we hash that out before going in deeper?

File: src/scripts/services/notifications/dimItemNotificationGenerator.factory.js (hunk 1 above)
Comment (EffortlessFury) [inline src/scripts/services/notifications/dimItemNotificationGenerator.factory.js]: The shuffle exists to make demoing the prototype a little easier and a little more variable. When you receive the notification, it notes that you received "{Item name} and x other items." The item chosen for the notification is the first in the list, allowing you to see a different image on each app load.

The "new" field was my first choice, but this "new" value does not clear on each inventory refresh; an item is considered new if it was obtained during the current session. Therefore, you'll be notified about the same items on each inventory refresh.

**Comment by bhollis** [comment]
@EffortlessFury if you don't mind, I'd like to close this PR because I don't think we have a path to merging it in. Let's chat about the goals of the system and how we could make it work with a new saved search feature instead.

**Comment by bhollis** [comment]
Linking #1874 which is how I think I'd like this to look.

**Pull request closed without merging**